test/testdata/rewriter/prop.rb

```diff
@@ -39,7 +39,7 @@ class AdvancedODM
     prop :t_nilable, T.nilable(String)
 
     prop :type, type: String
-    prop :object
+    prop :object  # error: Method `prop` does not exist on `T.class_of(AdvancedODM)`
```

Comment:
Why does this turn into a missing method error? Would it be possible to catch this earlier and give a less misleading error?

Reply:
In this case, that's actually not a misleading error, it's a correct one: the fact that the other ones _work_ is misleading, because if you look at the whole test file you'll see that `AdvancedODM` does not have an implementation listed for `prop` or `const`. If [you try to typecheck](https://sorbet.run/#%23%20typed%3A%20true%0Aclass%20A%0A%20%20prop%20%3Afoo%2C%20Integer%0Aend%0AT.reveal_type(A.new.foo)) this:
```ruby
class A
  prop :foo, Integer
end
T.reveal_type(A.new.foo)
```
it works fine, but if I run it, I get
```
samp.rb:6:in `<class:A>': undefined method `prop' for A:Class (NoMethodError)
```
The issue is that we desugar `prop` and `const` regardless of whether there's an underlying implementation for them in the class where they're desugared. This previously worked because the desugaring removes any call to `prop`, which means we don't bother resolving whether `prop` actually exists or not, and because my change _doesn't_ desugar `prop` if it doesn't have a type, then it sticks around in the tree and we (correctly!) observe that it doesn't exist on this class.

Reply:
I think I was just confused about the context that the error was being raised in, and it makes sense to me why it shows up as it does now. Thanks for explaining it!